packages/snaps-controllers/src/insights/SnapInsightsController.ts

```diff
@@ -0,0 +1,378 @@
+      });
+
+      this.#handleSnapRequest({
+        snapId,
+        handler: HandlerType.OnSignature,
+        params: { signature: payload, signatureOrigin },
+      })
+        .then((response) =>
+          this.#handleSnapResponse({
+            id,
+            snapId,
+            response: response as Record<string, Json>,
+          }),
+        )
+        .catch((error) => this.#handleSnapResponse({ id, snapId, error }));
+    });
+  }
+
+  /**
+   * Handle the transactionStatusUpdated event emitted by the TransactionController.
+   * This function will remove insights for the transaction in question
+   * once the transaction status has changed from unapproved.
+   *
+   * @param args - An options bag.
+   * @param args.transactionMeta - The transaction.
+   */
+  #handleTransactionStatusUpdate({
+    transactionMeta,
+  }: {
+    transactionMeta: TransactionMeta;
+  }) {
+    if (transactionMeta.status !== 'unapproved') {
+      this.#handleInsightCleanup(transactionMeta.id);
+    }
+  }
+
+  #handleInsightCleanup(id: string) {
+    if (!this.#hasInsight(id)) {
+      return;
+    }
+
+    // Delete interfaces from interface controller.
+    Object.values(this.state.insights[id])
+      .filter((insight) => insight.interfaceId)
+      .forEach((insight) => {
+        this.messagingSystem.call(
+          'SnapInterfaceController:deleteInterface',
+          insight.interfaceId as string,
+        );
+      });
+
+    this.update((state) => {
+      delete state.insights[id];
+    });
+  }
+
+  /**
+   * Handle sending a request to a given Snap with a given payload.
+   *
+   * @param args - An options bag.
+   * @param args.snapId - The Snap ID.
+   * @param args.handler - The handler to invoke.
+   * @param args.params - The JSON-RPC params to send.
+   * @returns The response from the Snap.
+   */
+  async #handleSnapRequest({
+    snapId,
+    handler,
+    params,
+  }: {
+    snapId: SnapId;
+    handler: HandlerType.OnTransaction | HandlerType.OnSignature;
+    params: Record<string, Json>;
+  }) {
+    return this.messagingSystem.call('SnapController:handleRequest', {
+      snapId,
+      origin: '',
+      handler,
+      request: {
+        method: '',
+        params,
+      },
+    });
+  }
+
+  /**
+   * Handle response from a given Snap by persisting the response or error in state.
+   *
+   * @param args - An options bag.
+   * @param args.id - The transaction or signature ID.
+   * @param args.snapId - The Snap ID.
+   * @param args.response - An optional response object returned by the Snap.
+   * @param args.error - An optional error returned by the Snap.
+   */
+  #handleSnapResponse({
+    id,
+    snapId,
+    response,
+    error,
+  }: {
+    id: string;
+    snapId: SnapId;
+    response?: Record<string, Json>;
+    error?: Error;
+  }) {
+    this.update((state) => {
+      state.insights[id][snapId].loading = false;
+      state.insights[id][snapId].interfaceId = response?.id as string;
```

Comment:
Why do we need a type cast here?

Reply:
The response type is typed as `Json`, but we know the return type for an insight Snap will contain `id` as a string or undefined.